source/game/field/KColData.cc

```diff
@@ -0,0 +1,353 @@
+    m_bbox.mMax.set(-9999999.0f);
+    m_bbox.mMin.set(9999999.0f);
+
+    for (size_t i = 0; i < m_prismCount; i++) {
+        const KCollisionPrism prism = getPrism(i);
+
+        const EGG::Vector3f fnrm = getNrm(prism.fnrm_i);
+        const EGG::Vector3f enrm1 = getNrm(prism.enrm1_i);
+        const EGG::Vector3f enrm2 = getNrm(prism.enrm2_i);
+        const EGG::Vector3f enrm3 = getNrm(prism.enrm3_i);
+        const EGG::Vector3f vtx1 = getPos(prism.pos_i);
+
+        const EGG::Vector3f vtx3 = GetVertex(prism.height, vtx1, fnrm, enrm3, enrm1);
+        const EGG::Vector3f vtx2 = GetVertex(prism.height, vtx1, fnrm, enrm3, enrm2);
+
+        m_bbox.mMin = m_bbox.mMin.minimize(vtx1);
+        m_bbox.mMin = m_bbox.mMin.minimize(vtx2);
+        m_bbox.mMin = m_bbox.mMin.minimize(vtx3);
+        m_bbox.mMax = m_bbox.mMax.maximize(vtx1);
+        m_bbox.mMax = m_bbox.mMax.maximize(vtx2);
+        m_bbox.mMax = m_bbox.mMax.maximize(vtx3);
+    }
+}
+
+EGG::Vector3f KColData::GetVertex(f32 height, const EGG::Vector3f &vertex1,
+        const EGG::Vector3f &fnrm, const EGG::Vector3f &enrm3, const EGG::Vector3f &enrm) {
+    EGG::Vector3f cross = fnrm.cross(enrm);
+    f32 dp = cross.ps_dot(enrm3);
+    cross *= (height / dp);
+
+    return cross + vertex1;
+}
+
+void KColData::lookupSphere(f32 radius, const EGG::Vector3f &pos, const EGG::Vector3f &prevPos,
+        KCLTypeMask typeMask) {
+    this->m_prismIter = searchBlock(pos);
+    this->m_pos = pos;
+    this->m_prevPos = prevPos;
+    this->m_movement = pos - prevPos;
+    this->m_radius = std::min(radius, m_sphereRadius);
+    this->m_typeMask = typeMask;
+}
+
+const u16 *KColData::searchBlock(const EGG::Vector3f &point) {
+    // Calculate the x, y, and z offsets of the point from the minimum
+    // corner of the tree's bounding box.
+    const int x = point.x - this->m_areaMinPos.x;
+    const int y = point.y - this->m_areaMinPos.y;
+    const int z = point.z - this->m_areaMinPos.z;
+
+    // Check if the point is outside the tree's bounding box in the x, y,
+    // or z dimensions. If it is, return 0.
+    if (x & this->m_areaXWidthMask || y & this->m_areaYWidthMask || z & this->m_areaZWidthMask) {
+        return nullptr;
+    }
+
+    // Initialize the current tree node to the root node of the tree.
+    u32 shift = this->m_blockWidthShift;
+    const u8 *curBlock = reinterpret_cast<const u8 *>(this->m_blockData);
+    s32 offset;
+
+    // Traverse the tree to find the leaf node containing the input point.
+    u32 index = 4 *
+            (((u32)z >> shift) << this->m_areaXYBlocksShift |
+                    ((u32)y >> shift) << this->m_areaXBlocksShift | (u32)x >> shift);
+
+    while (true) {
+        // Get the offset of the current node's child node.
+        offset = parse<u32>(*reinterpret_cast<const u32 *>(curBlock + index));
+
+        // If the offset is negative, the current node is a leaf node.
+        if ((offset & 0x80000000) != 0) {
+            break;
+        }
```

Comment:
Could we revise this so that the condition is checked before entering the loop? I don't know if that would be cleaner or not